File: platform/src/test/java/ru/abondin/hreasy/platform/repo/TestDataGenerator.java
package ru.abondin.hreasy.platform.repo;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import ru.abondin.hreasy.platform.repo.dict.*;
import ru.abondin.hreasy.platform.repo.employee.EmployeeCreationEntry;
import ru.abondin.hreasy.platform.repo.employee.EmployeeEntry;
import ru.abondin.hreasy.platform.repo.employee.EmployeeRepo;

import java.time.LocalDate;
import java.util.Arrays;

@Component()
@RequiredArgsConstructor
@Slf4j
public class TestDataGenerator {
    private final DepartmentRepo departmentRepo;
    private final DictPositionRepo positionRepo;
    private final DictLevelRepo levelRepo;
    private final EmployeeRepo employeeRepo;

    private DepartmentEntry devDep = new DepartmentEntry(null, "Development");
    private DepartmentEntry qaDep = new DepartmentEntry(null, "QA");

    private DictPositionEntry backendDeveloperPosition = new DictPositionEntry(null, "Backend Developer");
    private DictPositionEntry frontendDeveloperPosition = new DictPositionEntry(null, "Frontend Developer");
    private DictPositionEntry manualTesterPosition = new DictPositionEntry(null, "Manual Tester");

    private DictLevelEntry juniorLevel = new DictLevelEntry(null, "Junior", 10);
    private DictLevelEntry seniorLevel = new DictLevelEntry(null, "Senior", 100);

    EmployeeCreationEntry ivanovSeniorBackend = new EmployeeCreationEntry();
    EmployeeCreationEntry petrovJuniorBackend = new EmployeeCreationEntry();
    EmployeeCreationEntry sidorovSeniorFrontend = new EmployeeCreationEntry();
    EmployeeCreationEntry alekseevFiredSeniorFrontend = new EmployeeCreationEntry();
    EmployeeCreationEntry chackSeniorManualTester = new EmployeeCreationEntry();
    EmployeeCreationEntry admin = new EmployeeCreationEntry();

    public Mono generate() {
        return Mono.zip(generateDeps(), generatePositions(), generateLevels()).flatMap(it -> generateEmployes().collectList());
    }

    public Mono generateDeps() {
        return Mono.zip(saveDep(devDep), saveDep(qaDep));
    }

    public Mono generatePositions() {
        return Mono.zip(savePosition(backendDeveloperPosition)
                , savePosition(frontendDeveloperPosition)
                , savePosition(manualTesterPosition));
    }


    public Mono generateLevels() {
        return Mono.zip(saveLevel(juniorLevel), saveLevel(seniorLevel));
    }

    public Flux<EmployeeEntry> generateEmployes() {
        ivanovSeniorBackend.setDepartmentId(devDep.getId());
        ivanovSeniorBackend.setLevelId(seniorLevel.getId());
        ivanovSeniorBackend.setPositionId(backendDeveloperPosition.getId());
        ivanovSeniorBackend.setLastname("Ivanov");
        ivanovSeniorBackend.setFirstname("Ivan");
        ivanovSeniorBackend.setPatronymicName("Ivanovich");
        withEmail(ivanovSeniorBackend);

        petrovJuniorBackend.setDepartmentId(devDep.getId());
        petrovJuniorBackend.setLevelId(juniorLevel.getId());
        petrovJuniorBackend.setPositionId(backendDeveloperPosition.getId());
        petrovJuniorBackend.setLastname("Petrov");
        petrovJuniorBackend.setFirstname("Petr");
        petrovJuniorBackend.setPatronymicName("Petrovich");
        withEmail(petrovJuniorBackend);


        sidorovSeniorFrontend.setDepartmentId(devDep.getId());
        sidorovSeniorFrontend.setLevelId(seniorLevel.getId());
        sidorovSeniorFrontend.setPositionId(backendDeveloperPosition.getId());
        sidorovSeniorFrontend.setLastname("Sidorov");
        sidorovSeniorFrontend.setFirstname("Sidr");
        sidorovSeniorFrontend.setPatronymicName(null);
        withEmail(sidorovSeniorFrontend);

        alekseevFiredSeniorFrontend.setDepartmentId(devDep.getId());
        alekseevFiredSeniorFrontend.setLevelId(seniorLevel.getId());
        alekseevFiredSeniorFrontend.setPositionId(backendDeveloperPosition.getId());
        alekseevFiredSeniorFrontend.setLastname("Alekseev");
        alekseevFiredSeniorFrontend.setFirstname("Aleksey");
        alekseevFiredSeniorFrontend.setPatronymicName(null);
        alekseevFiredSeniorFrontend.setDateOfDismissal(LocalDate.of(2019, 12, 31));
        withEmail(alekseevFiredSeniorFrontend);


        chackSeniorManualTester.setDepartmentId(qaDep.getId());
        chackSeniorManualTester.setLevelId(seniorLevel.getId());
        chackSeniorManualTester.setPositionId(manualTesterPosition.getId());
        chackSeniorManualTester.setLastname("Chack");
        chackSeniorManualTester.setFirstname("Chackles");
        chackSeniorManualTester.setPatronymicName("Chack");
        withEmail(chackSeniorManualTester);


        admin.setDepartmentId(devDep.getId());
        admin.setLevelId(seniorLevel.getId());
        admin.setPositionId(frontendDeveloperPosition.getId());
        admin.setLastname("Admin");
        admin.setFirstname("Admin");
        admin.setPatronymicName("Admin");
        withEmail(admin);
        return Flux.fromIterable(Arrays.asList(
                ivanovSeniorBackend
                , petrovJuniorBackend
                , sidorovSeniorFrontend
                , alekseevFiredSeniorFrontend
                , chackSeniorManualTester
                , admin)).flatMap(employee -> saveEmployee(employee));

    }

    private Mono<? extends EmployeeCreationEntry> saveEmployee(EmployeeCreationEntry employee) {
        return employeeRepo.findIdByEmail(employee.getEmail()).map(id -> {
            employee.setId(id);
            return employee;
        }).switchIfEmpty(Mono.defer(() -> {
            log.info("Creating new employee ${employee.email}");
            return employeeRepo.save(employee);
        }));
    }


    private EmployeeEntry withEmail(EmployeeEntry employeeEntry) {
        employeeEntry.setEmail(employeeEntry.getFirstname() + "." + employeeEntry.getLastname() + "@stm-labs.ru");
        return employeeEntry;
    }


    private Mono<DepartmentEntry> saveDep(DepartmentEntry department) {
        return departmentRepo.findByName(department.getName()).map(d -> {
            department.setId(d.getId());
            return department;
        }).switchIfEmpty(departmentRepo.save(department));
    }


    private Mono<DictPositionEntry> savePosition(DictPositionEntry entry) {
        return positionRepo.findByName(entry.getName()).map(e -> {
                    entry.setId(e.getId());
                    return entry;
                }
        ).switchIfEmpty(positionRepo.save(entry));
    }

    private Mono<DictLevelEntry> saveLevel(DictLevelEntry entry) {
        return levelRepo.findByName(entry.getName()).map(e -> {
                    entry.setId(e.getId());
                    return entry;
                }
        ).switchIfEmpty(levelRepo.save(entry));
    }
}

File: platform/src/test/kotlin/ru/abondin/hreasy/platform/repo/TestDataGenerator.kt
package ru.abondin.hreasy.platform.repo

import org.springframework.stereotype.Component
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import ru.abondin.hreasy.platform.logger
import java.time.LocalDate

@Component()
class TestDataGenerator
(private val departmentRepo: DepartmentRepo,
 private val positionRepo: DictPositionRepo,
 private val levelRepo: DictLevelRepo,
 private val employeeRepo: EmployeeRepo
) {

    private val devDep = DepartmentEntry(null, "Development");
    private val qaDep = DepartmentEntry(null, "QA");

    private val backendDeveloperPosition = DictPositionEntry(null, "Backend Developer");
    private val frontendDeveloperPosition = DictPositionEntry(null, "Frontend Developer");
    private val manualTesterPosition = DictPositionEntry(null, "Manual Tester");

    private val juniorLevel = DictLevelEntry(null, "Junior", 10);
    private val seniorLevel = DictLevelEntry(null, "Senior", 100);

    val ivanovSeniorBackend = EmployeeCreationEntry();
    val petrovJuniorBackend = EmployeeCreationEntry();
    val sidorovSeniorFrontend = EmployeeCreationEntry();
    val alekseevFiredSeniorFrontend = EmployeeCreationEntry();
    val chackSeniorManualTester = EmployeeCreationEntry();
    val admin = EmployeeCreationEntry();

    fun generate() = Mono.zip(generateDeps(), generatePositions(), generateLevels()).flatMap { it -> generateEmployes().collectList() };

    fun generateDeps() = Mono.zip(saveDep(devDep), saveDep(qaDep));

    fun generatePositions() = Mono.zip(savePosition(backendDeveloperPosition)
            , savePosition(frontendDeveloperPosition)
            , savePosition(manualTesterPosition));


    fun generateLevels() = Mono.zip(saveLevel(juniorLevel), saveLevel(seniorLevel));

    fun generateEmployes(): Flux<EmployeeEntry> {
        ivanovSeniorBackend.departmentId = devDep.id;
        ivanovSeniorBackend.levelId = seniorLevel.id;
        ivanovSeniorBackend.positionId = backendDeveloperPosition.id;
        ivanovSeniorBackend.lastname = "Ivanov";
        ivanovSeniorBackend.firstname = "Ivan";
        ivanovSeniorBackend.patronymicName = "Ivanovich";
        withEmail(ivanovSeniorBackend);

        petrovJuniorBackend.departmentId = devDep.id;
        petrovJuniorBackend.levelId = juniorLevel.id;
        petrovJuniorBackend.positionId = backendDeveloperPosition.id;
        petrovJuniorBackend.lastname = "Petrov";
        petrovJuniorBackend.firstname = "Petr";
        petrovJuniorBackend.patronymicName = "Petrovich";
        withEmail(petrovJuniorBackend);


        sidorovSeniorFrontend.departmentId = devDep.id;
        sidorovSeniorFrontend.levelId = seniorLevel.id;
        sidorovSeniorFrontend.positionId = backendDeveloperPosition.id;
        sidorovSeniorFrontend.lastname = "Sidorov";
        sidorovSeniorFrontend.firstname = "Sidr";
        sidorovSeniorFrontend.patronymicName = null;
        withEmail(sidorovSeniorFrontend);

        alekseevFiredSeniorFrontend.departmentId = devDep.id;
        alekseevFiredSeniorFrontend.levelId = seniorLevel.id;
        alekseevFiredSeniorFrontend.positionId = backendDeveloperPosition.id;
        alekseevFiredSeniorFrontend.lastname = "Alekseev";
        alekseevFiredSeniorFrontend.firstname = "Aleksey";
        alekseevFiredSeniorFrontend.patronymicName = null;
        alekseevFiredSeniorFrontend.dateOfDismissal = LocalDate.of(2019, 12, 31);
        withEmail(alekseevFiredSeniorFrontend);


        chackSeniorManualTester.departmentId = qaDep.id;
        chackSeniorManualTester.levelId = seniorLevel.id;
        chackSeniorManualTester.positionId = manualTesterPosition.id;
        chackSeniorManualTester.lastname = "Chack";
        chackSeniorManualTester.firstname = "Chackles";
        chackSeniorManualTester.patronymicName = "Chack";
        withEmail(chackSeniorManualTester);



        admin.departmentId = devDep.id;
        admin.levelId = seniorLevel.id;
        admin.positionId = frontendDeveloperPosition.id;
        admin.lastname = "Admin";
        admin.firstname = "Admin";
        admin.patronymicName = "Admin";
        withEmail(admin);
        return Flux.fromIterable(arrayListOf(
                ivanovSeniorBackend
                , petrovJuniorBackend
                , sidorovSeniorFrontend
                , alekseevFiredSeniorFrontend
                , chackSeniorManualTester
                , admin)).flatMap { employee -> saveEmployee(employee) };

    }

    private fun saveEmployee(employee: EmployeeCreationEntry): Mono<out EmployeeCreationEntry>? {
        return employeeRepo.findIdByEmail(employee.email!!).map { id ->
            employee.id = id;
            return@map employee;
        }.switchIfEmpty(Mono.defer {
            logger().info("Creating new employee ${employee.email}")
            employeeRepo.save(employee)
        });
    }

    fun assignRoles() {

    }


    private fun withEmail(employeeEntry: EmployeeEntry): EmployeeEntry {
        employeeEntry.email = "${employeeEntry.firstname}.${employeeEntry.lastname}@stm-labs.ru";
        return employeeEntry;
    }


    private fun saveDep(department: DepartmentEntry): Mono<DepartmentEntry> =
            departmentRepo.findByName(department.name).map { d ->
                department.id = d.id
                department;
            }
                    .switchIfEmpty(departmentRepo.save(department));


    private fun savePosition(entry: DictPositionEntry): Mono<DictPositionEntry> = positionRepo.findByName(entry.name).map { e ->
        entry.id = e.id
        entry;
    }.switchIfEmpty(positionRepo.save(entry));

    private fun saveLevel(entry: DictLevelEntry): Mono<DictLevelEntry> = levelRepo.findByName(entry.name).map { e ->
        entry.id = e.id
        entry;
    }.switchIfEmpty(levelRepo.save(entry));

}